tigroo31/its-client
# client: add a first client

**How to test:**

0- Run the the mosquitto server 

```
$ sudo systemctl start mosquitto
...
```

1- Subscribe to the its fake server:

```
$ mosquitto_sub -v -h 127.0.0.1 -p 1883 -t "#"
...
```

2- Run the copycat client:

```
cargo run --color=always --bin its-client -- -H 127.0.0.1 -P 1883 --mqtt-client-id copycat -t 5GCroCo/outQueue
    Finished dev [unoptimized + debuginfo] target(s) in 0.04s
     Running `target/debug/its-client -H 127.0.0.1 -P 1883 --mqtt-client-id copycat -t 5GCroCo/outQueue`
Log is written to log/its-client_rCURRENT.log
...
```

3- Publish 2 denm's:

```
$ mosquitto_pub -t 5GCroCo/outQueue/v2x/denm/ora_car_testcon1/1/2/0/2/2/2/0/2/3/1/0/3/1/1/2/1/0/0/0 -m '{"type": "denm", "origin": "self", "version": "1.0.1", "source_uuid": "ora_car_testcon1", "timestamp": 1621602008340, "message": {"protocol_version": 2, "station_id": 16146508, "management_container": {"action_id": {"originating_station_id": 13101, "sequence_number": 13115}, "detection_time": 548686813340, "reference_time": 548686813340, "event_position": {"latitude": 435084065, "longitude": 12092271, "altitude": 20392}, "validity_duration": 10, "station_type": 5}, "situation_container": {"event_type": {"cause": 94, "subcause": 0}}, "location_container": {"event_speed": 37, "event_position_heading": 1585, "traces": []}}}'
$ mosquitto_pub -t 5GCroCo/outQueue/v2x/denm/ora_car_testcon1/1/2/0/2/2/2/0/2/3/1/0/3/1/1/2/1/0/0/0 -m '{"type": "denm", "origin": "self", "version": "1.0.1", "source_uuid": "ora_car_testcon1", "timestamp": 1621602008340, "message": {"protocol_version": 2, "station_id": 16146508, "management_container": {"action_id": {"originating_station_id": 13101, "sequence_number": 13115}, "detection_time": 548686813340, "reference_time": 548686813340, "event_position": {"latitude": 435084065, "longitude": 12092271, "altitude": 20392}, "validity_duration": 10, "station_type": 5}, "situation_container": {"event_type": {"cause": 94, "subcause": 0}}, "location_container": {"event_speed": 37, "event_position_heading": 1585, "traces": []}}}'
```

4- Check you received 2 denm's and sent one:

```
copycat_41470 denm received_on ora_broker/5GCroCo/outQueue/v2x/denm/ora_car_testcon1 16146508/13101/13115/548686813340/548686813340 at 1627660468376
copycat_41470 denm received_on ora_broker/5GCroCo/outQueue/v2x/denm/ora_car_testcon1 16146508/13101/13115/548686813340/548686813340 at 1627660581296
copycat_41470 denm sent_on ora_broker/5GCroCo/inQueue/v2x/denm/copycat_41470 16177978/13101/13115/554745391963/548686813340 at 1627660586964
```

5- Publish 2 cam's:

```
$ mosquitto_pub -t 5GCroCo/outQueue/v2x/cam/ora_car_testcon1/1/2/0/2/2/2/0/2/3/1/0/3/1/1/2/1/0/0/0 -m '{"type":"cam","origin":"self","version":"1.2.0","source_uuid":"ora_car_4044","timestamp":1618591358923,"message":{"protocol_version":2,"station_id":1077952580,"generation_delta_time":31571,"basic_container":{"station_type":5,"reference_position":{"latitude":486230934,"longitude":22419064,"altitude":10000},"confidence":{"position_confidence_ellipse":{"semi_major_confidence":0,"semi_minor_confidence":0,"semi_major_orientation":0},"altitude":1}},"high_frequency_container":{"heading":2375,"speed":37,"drive_direction":0,"vehicle_length":47,"vehicle_width":24,"longitudinal_acceleration":1,"lateral_acceleration":0,"vertical_acceleration":0,"yaw_rate":-27,"lane_position":0,"curvature":0,"confidence":{"heading":2,"speed":3,"vehicle_length":1,"vehicle_width":1,"longitudinal_acceleration":5,"yaw_rate":6,"lane_position":1,"curvature":7}},"low_frequency_container":{"vehicle_role":0,"exterior_lights":"00000000","path_history":[{"path_position":{},"path_delta_time":1}]}}}'
$ mosquitto_pub -t 5GCroCo/outQueue/v2x/cam/ora_car_testcon1/1/2/0/2/2/2/0/2/3/1/0/3/1/1/2/1/0/0/0 -m '{"type":"cam","origin":"self","version":"1.2.0","source_uuid":"ora_car_4044","timestamp":1618591358923,"message":{"protocol_version":2,"station_id":1077952580,"generation_delta_time":31571,"basic_container":{"station_type":5,"reference_position":{"latitude":486230934,"longitude":22419064,"altitude":10000},"confidence":{"position_confidence_ellipse":{"semi_major_confidence":0,"semi_minor_confidence":0,"semi_major_orientation":0},"altitude":1}},"high_frequency_container":{"heading":2375,"speed":37,"drive_direction":0,"vehicle_length":47,"vehicle_width":24,"longitudinal_acceleration":1,"lateral_acceleration":0,"vertical_acceleration":0,"yaw_rate":-27,"lane_position":0,"curvature":0,"confidence":{"heading":2,"speed":3,"vehicle_length":1,"vehicle_width":1,"longitudinal_acceleration":5,"yaw_rate":6,"lane_position":1,"curvature":7}},"low_frequency_container":{"vehicle_role":0,"exterior_lights":"00000000","path_history":[{"path_position":{},"path_delta_time":1}]}}}'
```

6- Check you received 2 cam's and sent one:

```
copycat_41470 cam received_on ora_broker/5GCroCo/outQueue/v2x/cam/ora_car_4044 1077952580/31571 at 1627660822918
copycat_41470 cam received_on ora_broker/5GCroCo/outQueue/v2x/cam/ora_car_4044 1077952580/31571 at 1627660828977
copycat_41470 cam sent_on ora_broker/5GCroCo/inQueue/v2x/cam/copycat_41470 1077984050/31571 at 1627660835296
```
